### PR TITLE
8329644: Discuss expected visitor evolution patterns in javax.lang.model.util

### PR DESCRIPTION
**Pull request by jddarcy**
Provide more concrete discussion of how the preview visitors are expected to evolve for different categories of language changes.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329644](https://bugs.openjdk.org/browse/JDK-8329644): Discuss expected visitor evolution patterns in javax.lang.model.util (**Enhancement** - P4)

### Reviewers
 * [Pavel Rappo](https://openjdk.org/census#prappo) (@pavelrappo - **Reviewer**) ⚠️ Review applies to [26160b42](https://git.openjdk.org/jdk/pull/18804/files/26160b42b25139f5ba7b15bd1f171af7ca00d42b)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**) ⚠️ Review applies to [2bdbdf43](https://git.openjdk.org/jdk/pull/18804/files/2bdbdf43da743abf36afd03ea57ced3a37aed255)
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to [2bdbdf43](https://git.openjdk.org/jdk/pull/18804/files/2bdbdf43da743abf36afd03ea57ced3a37aed255)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18804/head:pull/18804` \
`$ git checkout pull/18804`

Update a local copy of the PR: \
`$ git checkout pull/18804` \
`$ git pull https://git.openjdk.org/jdk.git pull/18804/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18804`

View PR using the GUI difftool: \
`$ git pr show -t 18804`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18804.diff">https://git.openjdk.org/jdk/pull/18804.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18804#issuecomment-2060078507)

### CODE REVIEW

**Comment by jddarcy** [comment]
The intention of this PR is to should approximate code diffs that would be needed to add support for new language features in the javax.lang.model visitors. Since the escaped code may be difficult to read, I'll cut and paste the generated javadoc output below:

API Note:
    Expected visitor evolution Link icon
    As the Java programming language evolves, the visitor interfaces of the language model also evolve as do the concrete visitors in this package. A preview language feature in JDK N may have API elements added in the set of visitors for the preview language level.

    The remainder of this note will show two examples of the API changes in the model and visitors that can be added to support a language feature. The examples will uses additions to the elements portion of the language model, but the updates to visitors for types or annotation values would be analogous. Two distinct cases are:

        the preview language construct has a corresponding new modeling interface
        the preview language construct only triggers the introduction of a new kind without a new modeling interface. 

    Adding visitor support for a top-level language construct Link icon
    Consider a new language feature, preview feature 1, in JDK N. This feature has a top-level element interface to model it:
```

     package javax.lang.model.element;
     /**
      * Represents a preview feature 1.
      *
      * @since N
      */
      public interface PreviewFeature1Element extends Element {
      // Methods to retrieve information specific to the preview feature...
      }
     
```

    A new element kind would also be introduce to model such a feature:

```
      //  Sample diff of ElementKind.java
      +    /**
      +     * A preview feature 1.
      +     * @since N
      +     */
      +     PREVIEW_FEATURE_1,
     

    A default method is added to ElementVisitor to accommodate the new construct:

     //  Sample diff for ElementVisitor.java
      +    /**
      +     * Visits a preview feature 1.
      +     *
      +     * @implSpec The default implementation visits a {@code
      +     * PreviewFeature1Element} by calling {@code visitUnknown(e, p)}.
      +     *
      +     * @param e  the element to visit
      +     * @param p  a visitor-specified parameter
      +     * @return a visitor-specified result
      +     * @since N
      +     */
      +    default R visitPreviewFeature1(PreviewFeature1Element e, P p) {
      +        return visitUnknown(e, p);
      +    }
```
     

    Given the default method on the visitor interface, the preview visitor classes need to override this method and take an appropriate action for the visitor's semantics:

```
     //  Sample diff for AbstractElementVisitorPreview.java
     //  Re-abstract visitPreviewFeature1.
      +    /**
      +     * {@inheritDoc ElementVisitor}
      +     *
      +     * @implSpec Visits a {@code PreviewFeature1Element} in a manner
      +     * defined by a subclass.
      +     *
      +     * @param e {@inheritDoc ElementVisitor}
      +     * @param p {@inheritDoc ElementVisitor}
      +     * @return a visitor-specified result
      +     * @since N
      +     */
      +    @Override
      +    public abstract R visitPreviewFeature1(PreviewFeature1Element e, P p);

     //  Sample diff for ElementKindVisitorPreview.java
     //  Take the default action for a preview feature 1.
      +
      +    /**
      +     * {@inheritDoc ElementVisitor}
      +     *
      +     * @implSpec This implementation calls {@code defaultAction}.
      +     *
      +     * @param e {@inheritDoc ElementVisitor}
      +     * @param p {@inheritDoc ElementVisitor}
      +     * @return  the result of {@code defaultAction}
      +     * @since N
      +     */
      +    @Override
      +    public R visitPreviewFeature1(PreviewFeature1Element e, P p) {
      +        return defaultAction(e, p);
      +    }

     //  Sample diff for ElementScannerPreview.java
     //  Scan the enclosed elements of a preview feature 1.
      +
      +    /**
      +     * {@inheritDoc ElementVisitor}
      +     *
      +     * @implSpec This implementation scans the enclosed elements.
      +     *
      +     * @param e {@inheritDoc ElementVisitor}
      +     * @param p {@inheritDoc ElementVisitor}
      +     * @return  {@inheritDoc ElementScanner6}
      +     * @since N
      +     */
      +    @Override
      +    public R visitPreviewFeature1(PreviewFeature1Element e, P p) {
      +        return scan(e.getEnclosedElements(), p);
      +    }

     //  Sample diff for SimpleElementVisitorPreview.java
     //  Take the default action for a preview feature 1.
      +    /**
      +     * {@inheritDoc ElementVisitor}
      +     *
      +     * @implSpec Visits a {@code PreviewFeature1Element} by calling
      +     * defaultAction.
      +     *
      +     * @param e {@inheritDoc ElementVisitor}
      +     * @param p {@inheritDoc ElementVisitor}
      +     * @return  {@inheritDoc ElementVisitor}
      +     * @since N
      +     */
      +    @Override
      +    public R visitPreviewFeature1(PreviewFeature1Element e, P p) {
      +        return defaultAction(e, p);
      +    }
```
     

    When preview feature 1 exits preview in JDK (N+k), a set of visitors for language level (N+k) would be added with the methods operating over the feature pulled in from the preview visitors. Each preview visitor would then have its direct superclass changed to the new corresponding (N+k) visitor.
    Adding visitor support for a language construct that is a new kind of an existing construct Link icon
    Consider a new language feature, preview feature 2, in JDK N. This feature has a new element kind without a new top-level element interface needed to model it. Concretely, assume a preview feature 2 is a new kind of variable; the changes would be analogous if the feature were a new kind of executable instead or new kind of another existing top-level construct. In that case, the API changes are more limited:
```

      //  Sample diff for ElementKind.java
      +    /**
      +     * A preview feature 2.
      +     * @since N
      +     */
      +     PREVIEW_FEATURE_2,
      ...
      // Update existing methods as needed
           public boolean isVariable() {
               return switch(this) {
               case ENUM_CONSTANT, FIELD, PARAMETER,
                    LOCAL_VARIABLE, EXCEPTION_PARAMETER, RESOURCE_VARIABLE,
      -             BINDING_VARIABLE -> true;
      +            BINDING_VARIABLE, PREVIEW_FEATURE_2 -> true;
               default -> false;
               };
           }
```
     

    The kind visitors need support for the new variety of element:

```
     // Update visitVariable in ElementKindVisitor6:
            ...
            * @implSpec This implementation dispatches to the visit method for
            * the specific {@linkplain ElementKind kind} of variable, {@code
            * ENUM_CONSTANT}, {@code EXCEPTION_PARAMETER}, {@code FIELD},
      -     * {@code LOCAL_VARIABLE}, {@code PARAMETER}, or {@code RESOURCE_VARIABLE}.
      +     * {@code LOCAL_VARIABLE}, {@code PARAMETER}, or {@code RESOURCE_VARIABLE},
      +     * or {@code PREVIEW_FEATURE_2}.
            *
            * @param e {@inheritDoc ElementVisitor}
            * @param p {@inheritDoc ElementVisitor}
            * @return  the result of the kind-specific visit method
            */
            @Override
            public R visitVariable(VariableElement e, P p) {
            ...
               case BINDING_VARIABLE:
                   return visitVariableAsBindingVariable(e, p);
       
      +        case PREVIEW_FEATURE_2:
      +            return visitVariableAsPreviewFeature2(e, p);
      +
               default:
                   throw new AssertionError("Bad kind " + k + " for VariableElement" + e);
            ...
      +    /**
      +     * Visits a {@code PREVIEW_FEATURE_2} variable element.
      +     *
      +     * @implSpec This implementation calls {@code visitUnknown}.
      +     *
      +     * @param e the element to visit
      +     * @param p a visitor-specified parameter
      +     * @return  the result of {@code visitUnknown}
      +     *
      +     * @since N
      +     */
      +    public R visitVariableAsPreviewFeature2(VariableElement e, P p) {
      +        return visitUnknown(e, p);
      +    }
```
     

    The preview element kind visitor in turn overrides visitVariableAsPreviewFeature2:

```
     // Sample diff for ElementKindVisitorPreview:
      +    /**
      +     * {@inheritDoc ElementKindVisitor6}
      +     *
      +     * @implSpec This implementation calls {@code defaultAction}.
      +     *
      +     * @param e {@inheritDoc ElementKindVisitor6}
      +     * @param p {@inheritDoc ElementKindVisitor6}
      +     * @return  the result of {@code defaultAction}
      +     *
      +     * @since N
      +     */
      +    public R visitVariableAsPreviewFeature2(VariableElement e, P p) {
      +        return defaultAction(e, p);
      +    }
```
     

    As in the case where a new interface is introduced, when preview feature 2 exits preview in JDK (N+k), a set of visitors for language level (N+k) would be added with the methods operating over the new feature in the kind visitors pulled in from the preview visitors. Each preview visitor would then have its direct superclass changed to the new corresponding (N+k) visitor.

**Comment by jddarcy** [comment]
I didn't use javadoc snippets because for the purposes of this PR it is important to present javadoc comments to the reader. Likewise, I only use a pre HTML tag and not pre and code allow the HTML escapes to be used to represent "/" and "@".

**Comment by jddarcy** [comment]
The code diffs were extracted from change temporary made in a repo to add support for a preview feature 1 and preview feature 2.

The two paths of adding API support are consistent with past practice when API support was added for previous language features; see issues linked to in JDK-8329644 for prior examples.

**Review by pavelrappo — COMMENTED** [review]
I have a few inline comments and a dumb question. The question: what happens if a preview feature exits the preview stage without being promoted to the standard, i.e. the preview feature is dropped?

**Comment by lahodaj** [comment]
> > > I have a few inline comments and a dumb question. The question: what happens if a preview feature exits the preview stage without being promoted to the standard, i.e. the preview feature is dropped?
> > 
> > 
> > When the preview feature is dropped, I would assume the API elements related to it would be removed. The `Preview` help with that, as they stay, only the methods are removed. At least, this is what I would expect.
> > Overall, looks reasonable to me. I wonder if the text should say the newly-added API entry points should be marked as preview? I assume it may be difficult to talk explicitly about `@jdk.internal.javac.PreviewFeature` here, but maybe a general note that it should be ensured the entry points are recognized as preview? Or did I miss that? Thanks.
> 
> Your latter paragraph is why I asked my question in the first place. If you reflect on an element of a preview feature Y, one could say that you indirectly use that preview feature Y, and so all the terms and conditions for preview features apply. If Y then goes away, you weren't warned/reminded that your code will no longer compile, and that's bad.

We mark the preview-related entry points of the javax.lang.model and Trees API with `@PreviewFeature(..., reflective=true)`, and this then produces warnings like:

```
$ javac  /tmp/TestPreview.java
/tmp/TestPreview.java:3: warning: [preview] COMPONENT_LOCAL_VARIABLE is a reflective preview API and may be removed in a future release.
        System.err.println(javax.lang.model.element.ElementKind.COMPONENT_LOCAL_VARIABLE);
                                                               ^
1 warning
```

I.e. the potential removal is not without a warning. (The difference between "reflective" and non-"reflective" Preview API is that the reflective API can be used without `--enable-preview`.)

> 
> Let's wait for Joe to respond.

**Comment by pavelrappo** [comment]
Ah! I didn't notice that `*(Visitor|Scanner)*Preview` classes were annotated with `@PreviewFeature(feature=PreviewFeature.Feature.LANGUAGE_MODEL, reflective=true)` because the diffs _in_ the doc comment didn't contain them. Good to know this has been taken care of, thanks! But what about constants and methods this PR proposes to add to `ElementKind`, `ElementVisitor`, and the likes?

**Comment by pavelrappo** [comment]
> We mark the preview-related entry points of the javax.lang.model and Trees API with `@PreviewFeature(..., reflective=true)`, and this then produces warnings like:
> 
> ```
> $ javac  /tmp/TestPreview.java
> /tmp/TestPreview.java:3: warning: [preview] COMPONENT_LOCAL_VARIABLE is a reflective preview API and may be removed in a future release.
>         System.err.println(javax.lang.model.element.ElementKind.COMPONENT_LOCAL_VARIABLE);
>                                                                ^
> 1 warning
> ```
> 
> I.e. the potential removal is not without a warning. (The difference between "reflective" and non-"reflective" Preview API is that the reflective API can be used without `--enable-preview`.)

Ah! I didn't notice that `*(Visitor|Scanner)*Preview` classes were annotated with `@PreviewFeature(feature=PreviewFeature.Feature.LANGUAGE_MODEL, reflective=true)` because the diffs _in_ the doc comment didn't contain them. Good to know this has been taken care of, thanks! But what about constants and methods this PR proposes to add to `ElementKind`, `ElementVisitor`, and the likes?

**Comment by jddarcy** [comment]
> > We mark the preview-related entry points of the javax.lang.model and Trees API with `@PreviewFeature(..., reflective=true)`, and this then produces warnings like:
> > ```
> > $ javac  /tmp/TestPreview.java
> > /tmp/TestPreview.java:3: warning: [preview] COMPONENT_LOCAL_VARIABLE is a reflective preview API and may be removed in a future release.
> >         System.err.println(javax.lang.model.element.ElementKind.COMPONENT_LOCAL_VARIABLE);
> >                                                                ^
> > 1 warning
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I.e. the potential removal is not without a warning. (The difference between "reflective" and non-"reflective" Preview API is that the reflective API can be used without `--enable-preview`.)
> 
> Ah! I didn't notice that `*(Visitor|Scanner)*Preview` classes were annotated with `@PreviewFeature(feature=PreviewFeature.Feature.LANGUAGE_MODEL, reflective=true)` because the diffs _in_ the doc comment didn't contain them. Good to know this has been taken care of, thanks! But what about constants and methods this PR proposes to add to `ElementKind`, `ElementVisitor`, and the likes? How does that work?

I'll add some text in the start of this note to indicate:

Any new API elements are marked as reflective preview (without saying how this is done). And any existing methods that are updated to support the new feature are *not* marked as preview.

If the feature is eventually withdrawn instead of being added to the platform, any API elements associated with it are expected to be removed.

**Comment by jddarcy** [comment]
I added an `@see` link in AbstractElementVisitorPreview.java to the new section. Once we agree on wording/syntax, I'll add a corresponding link in all the preview visitor classes.

**Review by lahodaj — APPROVED** [review]
Looks reasonable to me.

File: src/java.compiler/share/classes/javax/lang/model/util/package-info.java [hunk 1]
```diff
@@ -35,6 +35,274 @@
...
+ * types or annotation values would be analogous.
+ *
+ * Two distinct cases are:
+ * <ul>
+ *
+ * <li>the preview language construct has a corresponding new modeling
+ * interface
+ *
+ * <li>the preview language construct only triggers the introduction
+ * of a new kind <em>without</em> a new modeling interface.
```
Comment (jonathan-gibbons) [inline src/java.compiler/share/classes/javax/lang/model/util/package-info.java:63]: There are many kinds of kind. Can you qualify which kind you mean here?

File: src/java.compiler/share/classes/javax/lang/model/util/package-info.java [hunk 2]
```diff
@@ -35,6 +35,274 @@
...
+ *
+ * The kind visitors need support for the new variety of element:
+ * <pre>
+ * // Update visitVariable in ElementKindVisitor6:
+ *        ...
+ *        * &commat;implSpec This implementation dispatches to the visit method for
+ *        * the specific {&commat;linkplain ElementKind kind} of variable, {&commat;code
+ *        * ENUM_CONSTANT}, {&commat;code EXCEPTION_PARAMETER}, {&commat;code FIELD},
+ *  -     * {&commat;code LOCAL_VARIABLE}, {&commat;code PARAMETER}, or {&commat;code RESOURCE_VARIABLE}.
+ *  +     * {&commat;code LOCAL_VARIABLE}, {&commat;code PARAMETER}, or {&commat;code RESOURCE_VARIABLE},
```
Comment (jonathan-gibbons) [inline src/java.compiler/share/classes/javax/lang/model/util/package-info.java:242]: delete `or` at this position in the list

**Review by jonathan-gibbons — APPROVED** [review]
Two minor editorial suggestions

File: src/java.compiler/share/classes/javax/lang/model/util/package-info.java (hunk 2 above)
Comment (jddarcy) [reply src/java.compiler/share/classes/javax/lang/model/util/package-info.java]: Fixed; thanks.

File: src/java.compiler/share/classes/javax/lang/model/util/package-info.java (hunk 1 above)
Comment (jddarcy) [reply src/java.compiler/share/classes/javax/lang/model/util/package-info.java]: Add some more explicit information; thanks for the suggestion.

**Review by pavelrappo — APPROVED** [review]
Looks good; thanks.

File: src/java.compiler/share/classes/javax/lang/model/util/AbstractAnnotationValueVisitorPreview.java [hunk 3]
```diff
@@ -40,6 +40,8 @@
  * @param <R> the return type of this visitor's methods
  * @param <P> the type of the additional parameter to this visitor's methods.
  *
+* @see <a href="package-summary.html#expectedEvolution">
```
Comment (jonathan-gibbons) [inline src/java.compiler/share/classes/javax/lang/model/util/AbstractAnnotationValueVisitorPreview.java:43]: FWIW, I would expect that you could write this as

```
@see package-name##id   description
```

**Review by jonathan-gibbons — APPROVED** [review]

File: src/java.compiler/share/classes/javax/lang/model/util/AbstractAnnotationValueVisitorPreview.java (hunk 3 above)
Comment (jonathan-gibbons) [reply src/java.compiler/share/classes/javax/lang/model/util/AbstractAnnotationValueVisitorPreview.java:43]: Yes, it needs to be an element name before the `##`

**Pull request closed without merging**